src/Disciplina/Core/Types.hs

```diff
@@ -62,15 +68,65 @@ newtype CourseId = CourseId
     { getCourseId :: Word32
     } deriving (Eq, Ord, Show, Num)
 
+-- | Assignment can be either regular of final
+data AssignmentType = Regular | CourseFinal
+    deriving (Eq, Show, Generic)
+
+-- | Assignment doesn't contain actual assignment contents - only hash of them.
+data Assignment = Assignment
+    { aCourseId   :: !CourseId
+    -- ^ Course this assignement belongs to
+    , aType       :: !AssignmentType
+    -- ^ Assignment type
+    , aAssignment :: !Text
+    -- ^ Hash of assignment content
+    } deriving (Eq, Show, Generic)
+
 -- | 'AssignmentId' is a hash of assignment contents,
 -- which are stored off-chain.
-data Assignment
-
 type AssignmentId = Hash Assignment
 
--- | Stub type for submissions. Submission transaction
--- doesn't contain actual submission contents - only hash of them.
-type Submission = Hash Raw
+-- | Submission can be digital or offline
+-- TODO, better description of how these differs
+data SubmissionType = Digital | Offline
+    deriving (Eq, Show, Generic)
+
+-- | Student submissions
+data Submission = Submission
+    { sStudentId  :: !StudentId
+    -- ^ Student who created this submission
+    , sType       :: !SubmissionType
+    -- ^ Submission type
+    , sAssignment :: !Assignment
+    -- ^ Assignment of this submission
+    } deriving (Eq, Show, Generic)
+
+-- | Type alias for Submission signature.
+type SubmissionSig = Signature Submission
```

Comment:
Better to have `Signature (Hash Submission)` (and `Hash Submission` is basically `SubmissionId`)